include/circt/Dialect/MSFT/MSFTConstructs.td

```diff
@@ -67,3 +67,48 @@ def ChannelOp: MSFTOp<"constructs.channel",
     $input $clk $sym_name `(` $defaultStages `)` attr-dict `:` type($input)
   }];
 }
+
+// Linear, pipelineable datapath.
+def LinearOp : MSFTOp<"hlc.linear", [
+    SingleBlockImplicitTerminator<"OutputOp">,
+    IsolatedFromAbove
+  ]> {
+  let summary = "Model of a linear datapath which can be arbitrarily pipelined";
+  let description = [{
+    Defines a feed-forward datapath which can be scheduled into a pipeline.
+    Due to the feed-forwardness, the inner region is NOT a graph region.
+    Internally, only combinational operations (`comb`, `msft`, `hw`) are allowed.
+    The operation expects exactly one i1-typed port to be tagged with a
+    `msft.clock` attribute.
+
+    Example:
+    ```mlir
+    msft.module @foo(%in0 : i32, %in1 : i32, %in2 : i32, %clk : i1) -> (out: i32) -> {
+      %0 = msft.hlc.linear(%a = %in0, %b = %in1, %c = %in2) clock %clk (i32, i32, i32) -> (i32) {
+        %0 = comb.mul %a, %b : i32
+        %1 = comb.add %0, %c : i32
+        msft.output %1 : i32
+      }
+    }
+    ```
+  }];
+
+  let arguments = (ins Variadic<AnyType>:$ins, I1:$clock);
+  let results = (outs Variadic<AnyType>:$outs);
+  let regions = (region SizedRegion<1>:$datapath);
+
+  let skipDefaultBuilders = 1;
+  let builders = [
+    OpBuilder<(ins "StringAttr":$name, "hw::ModulePortInfo":$ports)>
+  ];
+
+  let extraClassDeclaration = [{
+    Block *getBodyBlock() { return &datapath().front(); }
+
+    // Returns the functional type of this operation.
+    FunctionType getType();
```

Comment:
Does this make sense in this new formulation?

Reply:
I think so - one usage is here: https://github.com/llvm/circt/pull/3680/files#diff-b2e16aae5c31acfeb5fece6ddff8e6641df21aa084c7c46c07846f2d71889396R1051